Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/person/core/PersonCertificatesExtensions.kt

```diff
@@ -184,61 +183,75 @@ private fun Collection<CwaCovidCertificate>.rule9FindOldRaTest(
     .filter { Duration(it.rawCertificate.test.sampleCollectedAt, nowUtc) > Duration.standardHours(24) }
     .maxByOrNull { it.rawCertificate.test.sampleCollectedAt }
 
-@Suppress("ReturnCount")
+@Suppress("ReturnCount", "ComplexMethod")
 fun Collection<CwaCovidCertificate>.findHighestPriorityCertificate(
     nowUtc: Instant = Instant.now()
-): CwaCovidCertificate {
-    Timber.d("findHighestPriorityCertificate(nowUtc=%s): %s", nowUtc, this)
-
-    rule1FindRecentPcrCertificate(nowUtc)?.let {
-        Timber.d("Rule 1 match (PCR Test Certificate <= 48 hours): %s", it)
-        return it
+): CwaCovidCertificate = this
+    .also { Timber.v("findHighestPriorityCertificate(nowUtc=%s): %s", nowUtc, this) }
+    .run {
+        val valid = filter {
+            it.getState() is CwaCovidCertificate.State.Valid || it.getState() is CwaCovidCertificate.State.ExpiringSoon
+        }
+        val expired = filter { it.getState() is CwaCovidCertificate.State.Expired }
+        val invalid = filter { it.getState() is CwaCovidCertificate.State.Invalid }
+        if (this.size != (valid.size + expired.size + invalid.size)) {
```

Comment:
Wouldn't this be something for a unit test?

Reply:
I'm not sure how I could test this.
The idea is to have something that causes a crash if someone adds new states in the future.
I will check again if I can make it an exhaustive `when`.